pkg/node-servant/static-pod-upgrade/upgrade.go

```diff
@@ -0,0 +1,172 @@
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"time"
+
+	"k8s.io/klog/v2"
+
+	"github.com/openyurtio/openyurt/pkg/node-servant/static-pod-upgrade/util"
+)
+
+const (
+	DefaultUpgradeDir                   = "openyurtio-upgrade"
+	DefaultStaticPodRunningCheckTimeout = 2 * time.Minute
+
+	// TODO: use constant value of static-pod controller
+	OTA  = "ota"
+	Auto = "auto"
+)
+
+var (
+	DefaultConfigmapPath = "/data"
+	DefaultManifestPath  = "/etc/kubernetes/manifests"
+)
+
+type Controller struct {
+	// Name of static pod
+	name string
+	// Namespace of static pod
+	namespace string
+	// Manifest file name of static pod
+	manifest string
+	// The latest static pod hash
+	hash string
+	// Only support `OTA` and `Auto`
+	upgradeMode string
+
+	// Manifest path of static pod, default `/etc/kubernetes/manifests/manifestName.yaml`
+	manifestPath string
+	// The backup manifest path, default `/etc/kubernetes/manifests/openyurtio-upgrade/manifestName.bak`
+	bakManifestPath string
+	// Default is `/data/podName`
+	configMapDataPath string
+	// The latest manifest path, default `/etc/kubernetes/manifests/openyurtio-upgrade/manifestName.upgrade`
+	upgradeManifestPath string
+}
+
+func New(name, namespace, manifest, hash, mode string) (*Controller, error) {
+	ctrl := &Controller{
+		name:        name,
+		namespace:   namespace,
+		manifest:    manifest,
+		hash:        hash,
+		upgradeMode: mode,
+	}
+
+	ctrl.manifestPath = filepath.Join(DefaultManifestPath, util.WithYamlSuffix(ctrl.manifest))
+	ctrl.bakManifestPath = filepath.Join(DefaultManifestPath, DefaultUpgradeDir, util.WithBackupSuffix(ctrl.manifest))
+	ctrl.configMapDataPath = filepath.Join(DefaultConfigmapPath, ctrl.manifest)
+	ctrl.upgradeManifestPath = filepath.Join(DefaultManifestPath, DefaultUpgradeDir, util.WithUpgradeSuffix(ctrl.manifest))
+
+	return ctrl, nil
+}
+
+func (ctrl *Controller) Upgrade() error {
+	// 1. Check old manifest and the latest manifest exist
+	if err := ctrl.checkManifestFileExist(); err != nil {
+		return err
+	}
+	klog.Info("Check old manifest and new manifest files existence success")
+
+	// 2. prepare the latest manifest
+	if err := ctrl.prepareManifest(); err != nil {
+		return err
+	}
+	klog.Info("Prepare upgrade manifest success")
+
+	// 3. execute upgrade operations
+	switch ctrl.upgradeMode {
+	case Auto:
+		return ctrl.AutoUpgrade()
+	}
+
+	return nil
+}
+
+func (ctrl *Controller) AutoUpgrade() error {
+	// (1) Back up the old manifest in case of upgrade failure
+	if err := ctrl.backupManifest(); err != nil {
+		return err
+	}
+	klog.Info("Auto upgrade backupManifest success")
+
+	// (2) Replace manifest and kubelet will upgrade the static pod automatically
+	if err := ctrl.replaceManifest(); err != nil {
+		return err
+	}
+	klog.Info("Auto upgrade replaceManifest success")
+
+	// (3) Verify the new static pod is running
+	ok, err := ctrl.verify()
+	if err != nil {
+		return err
+	}
+	if !ok {
+		return fmt.Errorf("the latest static pod is not running")
```

Comment:
maybe we need to fall into the backup manifest if upgrade failed.

Reply:
done